daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/NilEmptyCombinatorParsers.scala

```diff
@@ -33,55 +27,23 @@ abstract class NilOrValueParser(ctxt: TermRuntimeData, nilParser: Parser, valueP
   override lazy val runtimeDependencies = Vector()
 
   def parse(pstate: PState): Unit = {
-    var mark: PState.Mark = pstate.mark("NilOrValueParser")
-    var markLeakCausedByException = false
 
-    try {
+    // This isn't technically a point of uncertainty. Nothing in the nilParser
+    // can discriminate this PoU. We are really just using this as a technique
+    // to be able to reset back to this point if the nil parser fails.
+    pstate.withPointOfUncertainty("NilOrValueParser", ctxt) { pou =>
+
       nilParser.parse1(pstate)
 
       if (pstate.processorStatus ne Success) {
-        pstate.reset(mark)
-        mark = null
+        // did not find a nil. reset back to the pou and try to parse the value
+        pstate.resetToPointOfUncertainty(pou)
         valueParser.parse1(pstate)
       } else {
-        pstate.discard(mark)
```

Comment:
Loving these huge cleanups.

Reply:
Yeah, I've always hated the way we handled these mark leaks, so much code to maintain and mutable variables. The macro makes this much cleaner and consistent among all uses.